### DISCUSSION TITLE
Adding an ejector component?

### DISCUSSION THREAD

**Discussion by timorichert** [oemof/tespy#545] [Ideas]
Dear all,

I have started having a look at TESPy (and I'm impressed!), and also the heatpumps repository, offering readily modelled heat pump topologies based on TESPy (https://github.com/jfreissmann/heatpumps). Many heat pumps process improvements (IHX, economizer, cascading, ...) are covered by this excellent combined work. I'm wondering if it would be possible to also model an ejector as a heat pump process improvement, allowing to recover some of the otherwise lost work thrown away in the throttling valve.

**Ejector Cycle example** (Source: Annex 58 High-Temperature Heat Pumps Task 1 – Technologies, Task Report, https://heatpumpingtechnologies.org/publications/annex-58-high-temperature-heat-pumps-task-1-technologies-task-report/)

![image](https://github.com/user-attachments/assets/615da7d7-193b-48ad-bba5-95bb19012578)

Here is an example of a Python code repo that models an ejector, and that could be used outright or offer guidance how it could be implemented: https://github.com/AdamBuruzs/simpy_ejector

Happy to hear your thoughts!

All the best,
Timo

P.S.: I had searched for ejector before writing, and hope I didn't miss a previous discussion!

**Comment by fwitte** [comment]
Hi @timorichert and thank you for reaching out!

There was no public discussion I am aware of, thank you for bringing it up. I actually discussed this topic already with two groups of people, one of them being the main authors to the heatpumps dashboard.

I will have to take a look on how the python package you linked here is modeling the component, and how to transfer that to tespy. There is quite a lot of things going on right now, so I cannot promise a quick implementation. If you would be willing to help out here please let me know. Maybe we can form a small task-force with the other interested people, looking at you @jfreissmann and @maltefritz. I could offer a little bit of guidance on how to implement new components to accelerate the first version. I will update you, once I find the time to look into that library.

Best

Francesco

**Reply by fwitte** [reply]
Unfortunately, I do not have a refprop license to test the software you linked.

Apart from that, this could also be interesting: https://github.com/CalebBell/thermo/issues/11

**Reply by timorichert** [reply]
Hi Francesco,

thanks for your prompt reply! Good to hear that this topic is already under discussion.

Your link looks promising as well. I guess calculation speed will also be relevant, since TESPy is rather quick, and an overly complex component might slow the process down. As for [simpy_ejector](https://github.com/AdamBuruzs/simpy_ejector), I have posted a comment on a related issue on implementing an interface with CoolProp as well: https://github.com/AdamBuruzs/simpy_ejector/issues/1.

Unfortunately, I cannot dive into this at the moment, but I'll see if I can tackle that in the near future.

All the best,
Timo

**Reply by jfreissmann** [reply]
Hi @timorichert, hi @fwitte,

thank you for bringing up the ejector component. @maltefritz and I view it as the biggest missing piece of [heatpumps](https://github.com/jfreissmann/heatpumps) as well and would love to help getting it to tespy and eventually heatpumps. In addition to the resources you two brought up, the C# library [portyanikhin/VCRC](https://github.com/portyanikhin/VCRC) also models ejectors (see e.g. [EjectorFlows.cs](https://github.com/portyanikhin/VCRC/blob/main/src/VCRC/Components/EjectorFlows.cs)). This could be another point to work from, but unfortunately there are no refereces to literature given that I was able to find.

**Reply by jfreissmann** [reply]
Furthermore, there is a paper with the title [Thermodynamic modelling and parameter determination of ejector for ejection refrigeration systems](https://doi.org/10.1016/j.ijrefrig.2016.12.005) that looks extensive but promising.

**Reply by fwitte** [reply]
The description of the model actually looks quite straight forward... Sound of speed is also available in CoolProp. One could give it a shot! Open questions could be the estimation of correct dimensions of the ejector parts, and the internal iteration in the simulation process description might create a little bit of overhead. It should be doable nevertheless :)

**Reply by fwitte** [reply]
I started a notebook over here: https://github.com/fwitte/ejector (only first two steps, and the second step has a bug, which I could not find very fast). You are all invited to have a look :). Later the model should be solved in one step with a large system of equations. Potentially, the ejector could be a subsystem in tespy using existing components.
